src/native/eventpipe/ds-dump-protocol.c

```diff
@@ -93,6 +93,71 @@ dump_protocol_helper_unknown_command (
 	return true;
 }
 
+static
+bool
+dump_protocol_generate_core_dump_response_flatten (
+	void *payload,
+	uint8_t **buffer,
+	uint16_t *size)
+{
+    DiagnosticsGenerateCoreDumpResponsePayload *response = (DiagnosticsGenerateCoreDumpResponsePayload*)payload;
+
+	EP_ASSERT (payload != NULL);
+	EP_ASSERT (buffer != NULL);
+	EP_ASSERT (*buffer != NULL);
+	EP_ASSERT (size != NULL);
+
+	bool success = true;
+
+	// int32_t error
+	memcpy (*buffer, &response->error, sizeof (response->error));
+	*buffer += sizeof (response->error);
+	*size -= sizeof (response->error);
+
+	// LPCWSTR error message
+	success &= ds_ipc_message_try_write_string_utf16_t (buffer, size, response->error_message);
+
+	// Assert we've used the whole buffer we were given
+	EP_ASSERT(*size == 0);
+
+    return success;
+}
+
+static
+void
+dump_protocol_generate_core_dump_response (
+	DiagnosticsIpcStream *stream,
+	ds_ipc_result_t error,
+    ep_char8_t * errorText)
+{
+    DiagnosticsGenerateCoreDumpResponsePayload payload;
+	DiagnosticsIpcMessage message;
+	ds_ipc_message_init (&message);
+
+    // Initialize payload
+    payload.error = error;
```

Comment:
Since DiagnosticsGenerateCoreDumpResponsePayload is a stack allocated object, it should have an init/fini function pair, then you can init it directly with error and string as part of function arguments. That will also better isolate the ownership of the string that needs to be freed when the struct goes out of scope. If the intent is not for the struct to be owner of the string, then you should still make that clear, by allocating it into a local, ep_char16_t *, that you pass as argument into the init function and then fini won't don't anything (empty implementation), see process_protocol_helper_get_process_info_2 for similar scenarios.

Reply:
I think it is beneficial to have explicit ownership in C code. It makes it easier to identify who is in charge of allocating and freeing the memory for that string. In this case, you can still have the string owned elsewhere, but making the init/fini pair makes it clear that this struct doesn't own the string.